packages/hardhat-core/src/internal/artifacts/mutable.ts

```diff
@@ -0,0 +1,325 @@
+    // partial JSON strings get out of scope sooner and hence can be reclaimed
+    // by the GC if needed.
+    const file = await fsPromises.open(buildInfoPath, "w");
+    try {
+      {
+        const withoutOutput = JSON.stringify({
+          ...buildInfo,
+          output: undefined,
+        });
+
+        // We write the JSON (without output) except the last }
+        await file.write(withoutOutput.slice(0, -1));
+      }
+
+      {
+        const outputWithoutSourcesAndContracts = JSON.stringify({
+          ...buildInfo.output,
+          sources: undefined,
+          contracts: undefined,
+        });
+
+        // We start writing the output
+        await file.write(',"output":');
+
+        // Write the output object except for the last }
+        await file.write(outputWithoutSourcesAndContracts.slice(0, -1));
+
+        // If there were other field apart from sources and contracts we need
+        // a comma
+        if (outputWithoutSourcesAndContracts.length > 2) {
+          await file.write(",");
+        }
+      }
+
+      // Writing the sources
+      await file.write('"sources":{');
+
+      let isFirst = true;
+      for (const [name, value] of Object.entries(
+        buildInfo.output.sources ?? {}
+      )) {
+        if (isFirst) {
+          isFirst = false;
+        } else {
+          await file.write(",");
+        }
+
+        await file.write(`${JSON.stringify(name)}:${JSON.stringify(value)}`);
+      }
+
+      // Close sources object
+      await file.write("}");
+
+      // Writing the contracts
+      await file.write(',"contracts":{');
+
+      isFirst = true;
+      for (const [name, value] of Object.entries(
+        buildInfo.output.contracts ?? {}
+      )) {
+        if (isFirst) {
+          isFirst = false;
+        } else {
+          await file.write(",");
+        }
+
+        await file.write(`${JSON.stringify(name)}:${JSON.stringify(value)}`);
+      }
+
+      // close contracts object
+      await file.write("}");
+      // close output object
+      await file.write("}");
+      // close build info object
+      await file.write("}");
+    } finally {
+      await file.close();
+    }
+
+    return buildInfoPath;
+  }
+
+  private static _createBuildInfo(
+    id: string,
+    solcVersion: string,
+    solcLongVersion: string,
+    input: CompilerInput,
+    output: CompilerOutput
+  ): BuildInfo {
+    return {
+      id,
+      _format: BUILD_INFO_FORMAT_VERSION,
+      solcVersion,
+      solcLongVersion,
+      input,
+      output,
+    };
+  }
+
+  private _createDebugFile(artifactPath: string, pathToBuildInfo: string) {
+    const relativePathToBuildInfo = path.relative(
+      path.dirname(artifactPath),
+      pathToBuildInfo
+    );
+
+    const debugFile: DebugFile = {
+      _format: DEBUG_FILE_FORMAT_VERSION,
+      buildInfo: relativePathToBuildInfo,
+    };
+
+    return debugFile;
+  }
+
+  private static _getBuildInfoName(
```

Comment:
We discussed on a call and decided to leave the current approach in place, since it does have the advantage of indicating to the reader that the method doesn't use any instance state data.